python3.8-soco-0.20 fails to build

**Describe the bug**
A beets dependency `python3.8-soco-0.20` fails to build in the latest unstable.

**To Reproduce**
Steps to reproduce the behavior:
1. Update nixpkgs input to `0aeba64fb26e4defa0842a942757144659c6e29f`
2. Run `nixos-rebuild switch`
3. Have `beets` in packages.

**Expected behavior**
Beats should install without problems.

**Screenshots**
```sh
error: --- Error ------------------------------------------------------------------------------------- nix
builder for '/nix/store/mgs9bh5daygy3j93kgdljw4bflinw8ii-python3.8-soco-0.20.drv' failed with exit code 1; last 10 log lines:
          URI = "x-sonos-http:track%3a3402413.mp3?sid=2&amp;flags=32&amp;sn=1"
  >       assert desc_from_uri(URI) == "SA_RINCON2311_12345678"
  E       AssertionError: assert 'RINCON_AssociatedZPUDN' == 'SA_RINCON2311_12345678'
  E         - SA_RINCON2311_12345678
  E         + RINCON_AssociatedZPUDN

  tests/test_musicservices.py:325: AssertionError
  =========================== short test summary info ============================
  SKIPPED [66] tests/conftest.py:31: use --ip and an ip address to run integration tests.
  ================== 1 failed, 206 passed, 66 skipped in 8.78s ===================
error: --- Error ------------------------------------------------------------------------------------- nix
1 dependencies of derivation '/nix/store/9093qjchgqr0p0pqc1j3wmxkzlqa8mwx-beets-unstable-2021-01-29.drv' failed to build
error: --- Error ------------------------------------------------------------------------------------- nix
1 dependencies of derivation '/nix/store/mqk5a61mmqai0cj95418jc07kldavpj5-beets-unstable-2021-01-29-fish-completions.drv' failed to build
error: --- Error ------------------------------------------------------------------------------------- nix
1 dependencies of derivation '/nix/store/l5bx7br3q840yj77dr7ixbafr1xdp9z7-home-manager-path.drv' failed to build
error: --- Error ------------------------------------------------------------------------------------- nix
1 dependencies of derivation '/nix/store/0n7ak8bvd6gxq2a6bghrq3w18psh82vv-home-manager-generation.drv' failed to build
error: --- Error ------------------------------------------------------------------------------------- nix
1 dependencies of derivation '/nix/store/qj3pzngywbs2vkd0109dgvvljqvm466d-user-environment.drv' failed to build
error: --- Error ------------------------------------------------------------------------------------- nix
1 dependencies of derivation '/nix/store/h799fgq7234fyhfzaqd1aq8lxr55yyc7-etc.drv' failed to build
error: --- Error ------------------------------------------------------------------------------------- nix
1 dependencies of derivation '/nix/store/645pmwngigqfy0v2sypk3spp194hnpbi-nixos-system-muspus-21.05.20210228.0aeba64.drv' failed to build
```

**Notify maintainers**
@lovesegfault

**Metadata**
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.11.0, NixOS, 21.05.20210224.f6b5bfd (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20201205_a5d85d0`
 - channels(pimeys): `"home-manager-20.09"`
 - channels(root): `"nixos-20.09.3343.8d82c865b41"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

--- a/pkgs/development/python-modules/soco/default.nix
+++ b/pkgs/development/python-modules/soco/default.nix
@@ -1,29 +1,76 @@
-{ lib, buildPythonPackage, fetchPypi, xmltodict, requests
-, toml
-
-# Test dependencies
-, pytest, pytestcov, coveralls, pylint, flake8, graphviz, mock, sphinx
+{ buildPythonPackage
+, fetchFromGitHub
+, fetchpatch
+, graphviz
+, ifaddr
+, isPy27
+, lib
+, mock
+, nix-update-script
+, pytestCheckHook
+, requests
+, requests-mock
+, sphinx
 , sphinx_rtd_theme
+, toml
+, xmltodict
 }:
 
 buildPythonPackage rec {
   pname = "soco";
-  version = "0.19";
+  version = "0.21.2";
+  disabled = isPy27;
 
-  src = fetchPypi {
-    inherit pname version;
-    sha256 = "0dgca286vhrabm4r4jj545k895z6w2c70ars06vrjhf9cpgg7qck";
+  # N.B. We fetch from GitHub because the PyPI tarball doesn't contain the
+  # required files to run the tests.
+  src = fetchFromGitHub {
+    owner = "SoCo";
+    repo = "SoCo";
+    rev = "v${version}";
+    sha256 = "sha256-CCgkzUkt9YqTJt9tPBLmYXW6ZuRoMDd7xahYmNXgfM0=";
   };
 
-  propagatedBuildInputs = [ xmltodict requests toml ];
+  patches = [(fetchpatch {
+    url = "https://patch-diff.githubusercontent.com/raw/SoCo/SoCo/pull/811.patch";
+    sha256 = "sha256-GBd74c8zc25ROO411SZ9TTa+bi8yXJaaOQqY9FM1qj4=";
+  })];
+
+  # N.B. These exist because:
+  # 1. Upstream's pinning isn't well maintained, leaving dependency versions no
+  #    longer in nixpkgs.
+  # 2. There is no benefit for us to be running linting and coverage tests.
+  postPatch = ''
+    sed -i "/black/d" ./requirements-dev.txt
+    sed -i "/coveralls/d" ./requirements-dev.txt
+    sed -i "/flake8/d" ./requirements-dev.txt
+    sed -i "/pylint/d" ./requirements-dev.txt
+    sed -i "/pytest-cov/d" ./requirements-dev.txt
+  '';
+
+  propagatedBuildInputs = [
+    ifaddr
+    requests
+    toml
+    xmltodict
+  ];
+
   checkInputs = [
-    pytest pytestcov coveralls pylint flake8 graphviz mock sphinx
+    pytestCheckHook
+    graphviz
+    mock
+    requests-mock
+    sphinx
     sphinx_rtd_theme
   ];
 
-  meta = {
+  passthru.updateScript = nix-update-script {
+    attrPath = "python3Packages.${pname}";
+  };
+
+  meta = with lib; {
     homepage = "http://python-soco.com/";
     description = "A CLI and library to control Sonos speakers";
-    license = lib.licenses.mit;
+    license = licenses.mit;
+    maintainers = with maintainers; [ lovesegfault ];
   };
 }

